recipes/protobuf/all/conanfile.py

```diff
@@ -177,6 +178,23 @@ def build(self):
         cmake = self._configure_cmake()
         cmake.build()
 
+    def _create_shim(self):
+        if not tools.is_apple_os(self.settings.os) or not self.options.shared:
+            return
+
+        with tools.chdir(os.path.join(self.package_folder, "bin")):
+            shutil.move("protoc", "protoc.orig")
+            protoc = textwrap.dedent("""
+            #!/usr/bin/env bash
+            bindir=$(dirname "${BASH_SOURCE[0]}")
+            rootdir=$(dirname "${bindir}")
+            libdir="${rootdir}/lib"
+            protoc="${bindir}/protoc.orig"
+            DYLD_LIBRARY_PATH=${libdir} ${protoc} "$@"
+            """)
```

Comment:
Using `BASH_SOURCE` makes this only work on bash.
I copied the working directory detection from the scons pr.

```suggestion
            protoc = textwrap.dedent("""\
                #!/bin/sh
                bindir="$(dirname "$(realpath "$0")")"
                rootdir=$(dirname "${bindir}")
                libdir="${rootdir}/lib"
                protoc="${bindir}/protoc.orig"
                DYLD_LIBRARY_PATH=${libdir} ${protoc} "$@"
            """)
```

Reply:
Is it okay to enclose `$@` with double quotes?

Reply:
well, I am not sure if `realpath` is portable, I need to check
I am not an expert in bash, so no idea about `$@` and `BASH_SOURCE`
these are technical details, I want to discuss more if we want to go with shim approach in general, or do we want to try some other way(s)

Reply:
If I remember correctly, @jgsogo coined an idea about shims already.
I think it would fix the Apple sip problem once and for all.

Would it be useful for conan to provide this functionality?

Reply:
I think it could be useful, for sure.
however, I am not sure how to make generic portable shims (yet).
conan's nature that packages are relocatable. you don't know where are dependencies are going to be installed.
so IDK how should we locate paths to the shared libraries of dependencies.
in this case we can't use relative paths.

Reply:
I envision adding shims using a generator.
Then, at install time, conan can generate shims for all tagged binaries of dependencies + add this folder to the `PATH` env.

for protoc, it would generate:
```
conan_install_folder/shim_generator/protoc
conan_install_folder/shim_generator/protoc/__info__/protoc.txt
```
The binaries would be looking for a file called `__info__/argv[0].txt`. and execute the wrapped binary as a child.
This file would be generated by the conan shim generator and should contain all info: path of binary, environment variables, extra arguments, ...

FYI, chocolatey uses ShimGen. But that is proprietary software.
It looks like a shim generator deserves a project on itself.

Reply:
> If I remember correctly, @jgsogo coined an idea about shims already. I think it would fix the Apple sip problem once and for all.
> 
> Would it be useful for conan to provide this functionality? These shims can then be generated at consume time.

In my POC I used `virtualenv` generator. The shim basically runs the `activate.sh` before running the executable. As a Conan built-in feature it is trivial to generate it, but from a recipe itself is quite difficult as @SSE4 mentions above because we cannot hardcode the PATH to the dependencies (using `$(conan config home)/data/.../` is probably something we don't want to consider... 🙄 )

Reply:
I am not sure how far solution should go, and how portable should be it, in general.
the following use-case I would expect to work, at very least:
- run tools during the build (e.g. `conan create` or `conan install` scenarios)
- run tools from the conan cache (e.g. activating a `virtualenv` and running binary inside it, or directly locating the executable)

the following is tentative:
- if tools are copied outside of the conan cache, e.g. via `imports`, `deploy` generator or by other means

that might be tricky to make it work, as in that case location of the tool is unpredictable, as well as locations of some shared dependencies (like zlib, openssl, or whatever).
in practice, for that use-case, we can document to run `install_name_tool` or similar script to handle your own layout properly. you know better where and how to install the things.

we may rely on `virtualenv` generator or similar environment variables, we just need to be sure that in proper context proper environment variables are set and they don't collide.